pkg/cidata/cidata.TEMPLATE.d/boot/00-reboot-if-required.sh

```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+set -eux
+
+# Check if cloud-init forgot to reboot_if_required
+# (only implemented for apt at the moment, not dnf)
+
+if command -v dnf >/dev/null 2>&1; then
+	# dnf-utils needs to be installed, for needs-restarting
+	if dnf needs-restarting -h >/dev/null 2>&1; then
```

Comment:
Is `-h` always supported? I only see these options:

```
dnf needs-restarting [-u] [-r] [-s]
```

Reply:
It is --help, I am just checking to see if it is there - because of the false-is-true return code <sigh>

Reply:
I think it should have been spelled as `dnf -h needs-restarting`, then it would be less cryptic.

Reply:
```console
[root@b144ba347084 /]# dnf -h needs-restarting
No such command: needs-restarting. Please use /usr/bin/dnf --help
It could be a DNF plugin command, try: "dnf install 'dnf-command(needs-restarting)'"
[root@b144ba347084 /]# dnf install 'dnf-command(needs-restarting)'
Fedora 39 - x86_64                                                                                                                     9.9 MB/s |  89 MB     00:09    
Fedora 39 openh264 (From Cisco) - x86_64                                                                                               1.6 kB/s | 2.5 kB     00:01    
Fedora 39 - x86_64 - Updates                                                                                                           6.5 MB/s |  28 MB     00:04    
No match for argument: dnf-command(needs-restarting)
Error: Unable to find a match: dnf-command(needs-restarting)
[root@b144ba347084 /]# dnf install dnf-utils
Last metadata expiration check: 0:00:17 ago on Sun Jan 14 09:08:48 2024.
Dependencies resolved.
=======================================================================================================================================================================
 Package                                           Architecture                    Version                                      Repository                        Size
=======================================================================================================================================================================
Installing:
 dnf-utils                                         noarch                          4.4.4-1.fc39                                 updates                           37 k
Installing dependencies:
 dbus-libs                                         x86_64                          1:1.14.10-1.fc39                             fedora                           156 k
 dnf-plugins-core                                  noarch                          4.4.4-1.fc39                                 updates                           38 k
 fonts-filesystem                                  noarch                          1:2.0.5-12.fc39                              fedora                           8.2 k
 js-jquery                                         noarch                          3.6.4-2.fc39                                 fedora                           172 k
 python3-dateutil                                  noarch                          1:2.8.2-10.fc39                              fedora                           355 k
 python3-dbus                                      x86_64                          1.3.2-4.fc39                                 fedora                           157 k
 python3-distro                                    noarch                          1.8.0-6.fc39                                 fedora                            49 k
 python3-dnf-plugins-core                          noarch                          4.4.4-1.fc39                                 updates                          317 k
 python3-six                                       noarch                          1.16.0-12.fc39                               fedora                            41 k
 python3-systemd                                   x86_64                          235-5.fc39                                   fedora                           107 k
 web-assets-filesystem                             noarch                          5-20.fc39                                    fedora                           7.9 k
Installing weak dependencies:
 python-systemd-doc                                x86_64                          235-5.fc39                                   fedora                            75 k

Transaction Summary
=======================================================================================================================================================================
Install  13 Packages

Total download size: 1.5 M
Installed size: 4.7 M
Is this ok [y/N]: y
Downloading Packages:
(1/13): fonts-filesystem-2.0.5-12.fc39.noarch.rpm                                                                                      112 kB/s | 8.2 kB     00:00    
(2/13): python-systemd-doc-235-5.fc39.x86_64.rpm                                                                                       942 kB/s |  75 kB     00:00    
(3/13): dbus-libs-1.14.10-1.fc39.x86_64.rpm                                                                                            806 kB/s | 156 kB     00:00    
(4/13): js-jquery-3.6.4-2.fc39.noarch.rpm                                                                                              873 kB/s | 172 kB     00:00    
(5/13): python3-distro-1.8.0-6.fc39.noarch.rpm                                                                                         1.0 MB/s |  49 kB     00:00    
(6/13): python3-dbus-1.3.2-4.fc39.x86_64.rpm                                                                                           2.2 MB/s | 157 kB     00:00    
(7/13): python3-dateutil-2.8.2-10.fc39.noarch.rpm                                                                                      2.4 MB/s | 355 kB     00:00    
(8/13): python3-six-1.16.0-12.fc39.noarch.rpm                                                                                          704 kB/s |  41 kB     00:00    
(9/13): python3-systemd-235-5.fc39.x86_64.rpm                                                                                          2.2 MB/s | 107 kB     00:00    
(10/13): web-assets-filesystem-5-20.fc39.noarch.rpm                                                                                    188 kB/s | 7.9 kB     00:00    
(11/13): dnf-plugins-core-4.4.4-1.fc39.noarch.rpm                                                                                      288 kB/s |  38 kB     00:00    
(12/13): dnf-utils-4.4.4-1.fc39.noarch.rpm                                                                                             256 kB/s |  37 kB     00:00    
(13/13): python3-dnf-plugins-core-4.4.4-1.fc39.noarch.rpm                                                                              1.1 MB/s | 317 kB     00:00    
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                  1.5 MB/s | 1.5 MB     00:01     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                               1/1 
  Installing       : python3-six-1.16.0-12.fc39.noarch                                                                                                            1/13 
  Installing       : python3-dateutil-1:2.8.2-10.fc39.noarch                                                                                                      2/13 
  Installing       : python3-distro-1.8.0-6.fc39.noarch                                                                                                           3/13 
  Installing       : fonts-filesystem-1:2.0.5-12.fc39.noarch                                                                                                      4/13 
  Installing       : web-assets-filesystem-5-20.fc39.noarch                                                                                                       5/13 
  Installing       : js-jquery-3.6.4-2.fc39.noarch                                                                                                                6/13 
  Installing       : python-systemd-doc-235-5.fc39.x86_64                                                                                                         7/13 
  Installing       : python3-systemd-235-5.fc39.x86_64                                                                                                            8/13 
  Installing       : dbus-libs-1:1.14.10-1.fc39.x86_64                                                                                                            9/13 
  Installing       : python3-dbus-1.3.2-4.fc39.x86_64                                                                                                            10/13 
  Installing       : python3-dnf-plugins-core-4.4.4-1.fc39.noarch                                                                                                11/13 
  Installing       : dnf-plugins-core-4.4.4-1.fc39.noarch                                                                                                        12/13 
  Installing       : dnf-utils-4.4.4-1.fc39.noarch                                                                                                               13/13 
  Running scriptlet: dnf-utils-4.4.4-1.fc39.noarch                                                                                                               13/13 
  Verifying        : dbus-libs-1:1.14.10-1.fc39.x86_64                                                                                                            1/13 
  Verifying        : fonts-filesystem-1:2.0.5-12.fc39.noarch                                                                                                      2/13 
  Verifying        : js-jquery-3.6.4-2.fc39.noarch                                                                                                                3/13 
  Verifying        : python-systemd-doc-235-5.fc39.x86_64                                                                                                         4/13 
  Verifying        : python3-dateutil-1:2.8.2-10.fc39.noarch                                                                                                      5/13 
  Verifying        : python3-dbus-1.3.2-4.fc39.x86_64                                                                                                             6/13 
  Verifying        : python3-distro-1.8.0-6.fc39.noarch                                                                                                           7/13 
  Verifying        : python3-six-1.16.0-12.fc39.noarch                                                                                                            8/13 
  Verifying        : python3-systemd-235-5.fc39.x86_64                                                                                                            9/13 
  Verifying        : web-assets-filesystem-5-20.fc39.noarch                                                                                                      10/13 
  Verifying        : dnf-plugins-core-4.4.4-1.fc39.noarch                                                                                                        11/13 
  Verifying        : dnf-utils-4.4.4-1.fc39.noarch                                                                                                               12/13 
  Verifying        : python3-dnf-plugins-core-4.4.4-1.fc39.noarch                                                                                                13/13 

Installed:
  dbus-libs-1:1.14.10-1.fc39.x86_64      dnf-plugins-core-4.4.4-1.fc39.noarch         dnf-utils-4.4.4-1.fc39.noarch           fonts-filesystem-1:2.0.5-12.fc39.noarch
  js-jquery-3.6.4-2.fc39.noarch          python-systemd-doc-235-5.fc39.x86_64         python3-dateutil-1:2.8.2-10.fc39.noarch python3-dbus-1.3.2-4.fc39.x86_64       
  python3-distro-1.8.0-6.fc39.noarch     python3-dnf-plugins-core-4.4.4-1.fc39.noarch python3-six-1.16.0-12.fc39.noarch       python3-systemd-235-5.fc39.x86_64      
  web-assets-filesystem-5-20.fc39.noarch

Complete!
[root@b144ba347084 /]# dnf -h needs-restarting
usage: dnf needs-restarting [-c [config file]] [-q] [-v] [--version] [--installroot [path]] [--nodocs] [--noplugins] [--enableplugin [plugin]]
                            [--disableplugin [plugin]] [--releasever RELEASEVER] [--setopt SETOPTS] [--skip-broken] [-h] [--allowerasing] [-b | --nobest] [-C]
                            [-R [minutes]] [-d [debug level]] [--debugsolver] [--showduplicates] [-e ERRORLEVEL] [--obsoletes] [--rpmverbosity [debug level name]]
                            [-y] [--assumeno] [--enablerepo [repo]] [--disablerepo [repo] | --repo [repo]] [--enable | --disable] [-x [package]]
                            [--disableexcludes [repo]] [--repofrompath [repo,path]] [--noautoremove] [--nogpgcheck] [--color COLOR] [--refresh] [-4] [-6]
                            [--destdir DESTDIR] [--downloadonly] [--comment COMMENT] [--bugfix] [--enhancement] [--newpackage] [--security] [--advisory ADVISORY]
                            [--bz BUGZILLA] [--cve CVES] [--sec-severity {Critical,Important,Moderate,Low}] [--forcearch ARCH] [-u] [-r] [-s]

determine updated binaries that need restarting

General DNF options:
  -c [config file], --config [config file]
                        config file location
  -q, --quiet           quiet operation
  -v, --verbose         verbose operation
  --version             show DNF version and exit
  --installroot [path]  set install root
  --nodocs              do not install documentations
  --noplugins           disable all plugins
  --enableplugin [plugin]
                        enable plugins by name
  --disableplugin [plugin]
                        disable plugins by name
  --releasever RELEASEVER
                        override the value of $releasever in config and repo files
  --setopt SETOPTS      set arbitrary config and repo options
  --skip-broken         resolve depsolve problems by skipping packages
  -h, --help, --help-cmd
                        show command help
  --allowerasing        allow erasing of installed packages to resolve dependencies
  -b, --best            try the best available package versions in transactions.
  --nobest              do not limit the transaction to the best candidate
  -C, --cacheonly       run entirely from system cache, don't update cache
  -R [minutes], --randomwait [minutes]
                        maximum command wait time
  -d [debug level], --debuglevel [debug level]
                        debugging output level
  --debugsolver         dumps detailed solving results into files
  --showduplicates      show duplicates, in repos, in list/search commands
  -e ERRORLEVEL, --errorlevel ERRORLEVEL
                        error output level
  --obsoletes           enables dnf's obsoletes processing logic for upgrade or display capabilities that the package obsoletes for info, list and repoquery
  --rpmverbosity [debug level name]
                        debugging output level for rpm
  -y, --assumeyes       automatically answer yes for all questions
  --assumeno            automatically answer no for all questions
  --enablerepo [repo]   Temporarily enable repositories for the purpose of the current dnf command. Accepts an id, a comma-separated list of ids, or a glob of ids.
                        This option can be specified multiple times.
  --disablerepo [repo]  Temporarily disable active repositories for the purpose of the current dnf command. Accepts an id, a comma-separated list of ids, or a glob
                        of ids. This option can be specified multiple times, but is mutually exclusive with `--repo`.
  --repo [repo], --repoid [repo]
                        enable just specific repositories by an id or a glob, can be specified multiple times
  --enable              enable repos with config-manager command (automatically saves)
  --disable             disable repos with config-manager command (automatically saves)
  -x [package], --exclude [package], --excludepkgs [package]
                        exclude packages by name or glob
  --disableexcludes [repo], --disableexcludepkgs [repo]
                        disable excludepkgs
  --repofrompath [repo,path]
                        label and path to an additional repository to use (same path as in a baseurl), can be specified multiple times.
  --noautoremove        disable removal of dependencies that are no longer used
  --nogpgcheck          disable gpg signature checking (if RPM policy allows)
  --color COLOR         control whether color is used
  --refresh             set metadata as expired before running the command
  -4                    resolve to IPv4 addresses only
  -6                    resolve to IPv6 addresses only
  --destdir DESTDIR, --downloaddir DESTDIR
                        set directory to copy packages to
  --downloadonly        only download packages
  --comment COMMENT     add a comment to transaction
  --bugfix              Include bugfix relevant packages, in updates
  --enhancement         Include enhancement relevant packages, in updates
  --newpackage          Include newpackage relevant packages, in updates
  --security            Include security relevant packages, in updates
  --advisory ADVISORY, --advisories ADVISORY
                        Include packages needed to fix the given advisory, in updates
  --bz BUGZILLA, --bzs BUGZILLA
                        Include packages needed to fix the given BZ, in updates
  --cve CVES, --cves CVES
                        Include packages needed to fix the given CVE, in updates
  --sec-severity {Critical,Important,Moderate,Low}, --secseverity {Critical,Important,Moderate,Low}
                        Include security relevant packages matching the severity, in updates
  --forcearch ARCH      Force the use of an architecture

Needs-restarting command-specific options:
  -u, --useronly        only consider this user's processes
  -r, --reboothint      only report whether a reboot is required (exit code 1) or not (exit code 0)
  -s, --services        only report affected systemd services
```

It is installed by default in Fedora, but there could be custom templates using DNF that doesn't have it.

And of course the help text is broken / doesn't work, but that is to be expected (nobody tests these...)

Reply:
I wonder if we shouldn't attempt to install the plugin (but only when `package_reboot_if_required` is true)? Something like:

```shell
if command -v dnf >/dev/null 2>&1; then
    if grep -q "package_reboot_if_required.*true" "${LIMA_CIDATA_MNT}/user-data"; then
        # dnf-utils needs to be installed, for needs-restarting
 	    if ! dnf -h needs-restarting >/dev/null 2>&1; then
            dnf install dnf-utils
        fi
    fi
    if dnf -h needs-restarting >/dev/null 2>&1; then
    	# needs-restarting returns "false" if needed (!)
 		if ! dnf needs-restarting -r >/dev/null 2>&1; then
 			systemctl reboot
 		fi
 	fi
fi
```

Reply:
The best would to propose a PR to cloud-init (to fix the issue), but that's going to take longer